dhcpv4/client.go

```diff
@@ -194,17 +195,36 @@ func BroadcastSendReceive(sendFd, recvFd int, packet *DHCPv4, readTimeout, write
 		defer conn.Close()
 		conn.SetReadDeadline(time.Now().Add(readTimeout))
 
-		buf := make([]byte, MaxUDPReceivedPacketSize)
-		n, _, _, _, err := conn.(*net.UDPConn).ReadMsgUDP(buf, []byte{})
-		if err != nil {
-			errs <- err
-			return
-		}
-
-		response, err = FromBytes(buf[:n])
-		if err != nil {
-			errs <- err
-			return
+		for {
+			buf := make([]byte, MaxUDPReceivedPacketSize)
+			n, _, _, _, err := conn.(*net.UDPConn).ReadMsgUDP(buf, []byte{})
+			if err != nil {
+				errs <- err
+				return
+			}
+
+			response, err = FromBytes(buf[:n])
+			if err != nil {
+				errs <- err
+				return
+			}
+			// check that this is a response to our message
+			if response.TransactionID() != packet.TransactionID() {
+				continue
+			}
+			// wait for a response message
+			if response.Opcode() != OpcodeBootReply {
+				continue
+			}
+			// if we are not requested to wait for a specific message type,
+			// return what we have
+			if messageType == MessageTypeNone {
+				break
+			}
+			// break if it's a reply of the desired type, continue otherwise
+			if response.MessageType() != nil && *response.MessageType() == messageType {
```

Comment:
I think this should be
```
if response != nil && *response.MessageType() == messgeType
```

Reply:
Oh never mind, I just didn't read down far enough. LGTM!

Reply:
I can do the check above, as response is already used multiple times before this point